Modules/readline.c

```diff
@@ -147,8 +157,21 @@ readline_free(void *m)
 
 static PyModuleDef readlinemodule;
 
-#define readlinestate_global ((readlinestate *)PyModule_GetState(PyState_FindModule(&readlinemodule)))
+static inline readlinestate*
+get_global_readline_state(bool set_error)
+{
+    PyObject *mod = PyState_FindModule(&readlinemodule);
+    if (mod == NULL){
+        if (set_error && !PyErr_Occurred()){
+            PyErr_SetString(PyExc_RuntimeError,
+                            "readline module state not initialized");
+        }
+        return NULL;
+    }
+    return get_readline_state(mod);
+}
 
+#define get_global_readline_state_with_error() get_global_readline_state(true)
```

Comment:
I'd say always return an error.  The caller can clear it (`PyErr_Clear()`) if desired.

```suggestion
static inline readlinestate*
get_global_readline_state(void)
{
    PyObject *mod = PyState_FindModule(&readlinemodule);
    if (mod == NULL){
        if (!PyErr_Occurred()){
            PyErr_SetString(PyExc_RuntimeError,
                            "readline module state not initialized");
        }
        return NULL;
    }
    Py_INCREF(mod);
    readlinestate *state = get_readline_state(mod);
    Py_DECREF(mod);
    return state;
}
```

Reply:
Also, I'd recommend naming the function `get_module_state`.  (If there were any ambiguity then I'd call it `get_current_module_state()`, but I don't think there is any.)

While we're at it, I'd also rename `get_readline_state()` to `get_module_state_from_mod()` or something like that, to reduce any confusion.

Reply:
Hmm, it seems like all the uses of `get_global_readline_state_with_error()` can be replaced with `get_readline_state()`.  Thus "set_error" would always be `false`.  Each of those cases would happen in a hook, where we always throw away an exceptions, so consider being explicit about that here:

```suggestion
static inline readlinestate*
get_hook_module_state(void)
{
    PyObject *mod = PyState_FindModule(&readlinemodule);
    if (mod == NULL){
        /* The hook is always going to fail(?), so remove it (and all others). */
        readline_cleanup();
        PyErr_Clear();
        return NULL;
    }
    Py_INCREF(mod);
    readlinestate *state = get_readline_state(mod);
    Py_DECREF(mod);
    return state;
}
```

Furthremore, this function would probably best list right after `on_hook()`.

Arguably, in that `NULL` case we should always call `readline_cleanup()`.  We don't want broken hooks triggering over and over.  However, ...

A better thing would be to check the state of affairs in `setup_readline()`, *before* setting any hooks or otherwise configuring readline.  That would involve calling `PyState_FindModule()` to make sure it doesn't return `NULL`.  That way there would be no need to ever call `readline_cleanup()` from inside a hook. [^1]

[^1]: Would we use `readline_cleanup()` for anything else or could we drop it?  If this were a multi-phase init module then we'd certainly call it in the module's clear/free func.

Thus:

```suggestion
static inline readlinestate*
get_hook_module_state(void)
{
    PyObject *mod = PyState_FindModule(&readlinemodule);
    if (mod == NULL){
        /* We already made sure the module is findable in setup_readline(). */
        assert(PyErr_Occured());
        PyErr_Clear();
        return NULL;
    }
    Py_INCREF(mod);
    readlinestate *state = get_readline_state(mod);
    Py_DECREF(mod);
    return state;
}
```